### PR TITLE
fix(ember__test-helpers): specify context for getContext

### PR DESCRIPTION
**Pull request by NullVoxPopuli**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [ ] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If adding a new definition:
- [ ] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [ ] If this is for an NPM package, match the name. If not, do not conflict with the name of an NPM package.
- [ ] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [ ] Represents shape of module/library [correctly](https://www.typescriptlang.org/docs/handbook/declaration-files/library-structures.html)
- [ ] `tslint.json` should be present and it shouldn't have any additional or disabling of rules. Just content as `{ "extends": "dtslint/dt.json" }`. If for reason the some rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]`  and not for whole package so that the need for disabling can be reviewed.
- [ ] `tsconfig.json` should have `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

If changing an existing definition:
- [ ] Provide a URL to documentation or source code which provides context for the suggested changes: <<url here>>
 - official documentation is not writtent with the specifity that typescript needs for `getContext()` to be correct.
 - change tested here: https://github.com/NullVoxPopuli/emberclear/pull/882/commits/f61d9a63bf6c06cc4ee503825a3d37534801c39e
 - https://github.com/emberjs/ember-test-helpers/blob/a8f623d6710e55c1fee670f2be424385345fa347/addon-test-support/%40ember/test-helpers/setup-context.ts#L58
   - actually looks like there are types for this... but not exposed during build :-\
   - more to copy-paste from I guess :-\ 

- [ ] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.
- [x] Include [tests for your changes](https://github.com/DefinitelyTyped/DefinitelyTyped#testing)
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`. If for reason the any rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]` and not for whole package so that the need for disabling can be reviewed.

If removing a declaration:
- [ ] If a package was never on Definitely Typed, you don't need to do anything. (If you wrote a package and provided types, you don't need to register it with us.)
- [ ] Delete the package's directory.
- [ ] Add it to `notNeededPackages.json`.

### CODE REVIEW

**Comment by NullVoxPopuli** [comment]
I disagree with the failing lint rule. 
I don't think the good is good here:
```ts
// Bad:

function parse<T>(): T;
const x = parse<number>();

// Good:

function parse(): {};
const x = parse() as number;
```

**Review by chriskrycho — CHANGES_REQUESTED** [review]
Thanks for the PR!

1. Agreement with the lint rule here isn’t actually necessary. Following it is, though! :wink:
2. Please work with maintainers of `@ember/test-helpers` to actually get documented what (if anything) further should be publicly specified for the types here. We won’t merge types that aren’t supported by docs. The fact that internal types exist does *not* mean they’re intended to be public in their current form.

Once we have those docs, I will happily re-review!

**Comment by danger-public** [comment]
Inspecting the JavaScript source for this package found some properties that are not in the .d.ts files.
The check for missing properties isn't always right, so take this list as advice, not a requirement.
## ember__test-helpers ([<kbd>unpkg</kbd>](https://unpkg.com/browse/ember__test-helpers@latest/))
was missing the following properties:
1. name
1. init
1. included
1. treeForAddonTestSupport
1. treeForVendor
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 29c923313422b1de160595704633f72ae6791807
</p>

**Comment by NullVoxPopuli** [comment]
Thanks for looking @chriskrycho !

**Comment by elibarzilay** [comment]
@NullVoxPopuli: what's the status of this?

**Comment by NullVoxPopuli** [comment]
I think this is still valuable to have, but I need to sync with the @ember/test-helpers folks
@rwjblue, etc

**Comment by chriskrycho** [comment]
No syncing required: we're not shipping types from core Ember libraries at this point. (Soon™, hopefully!)

Sorry about the delay on reviewing, though; I'll review this later today!

**Comment by chriskrycho** [comment]
Though, @dfreeman pointed out to me that the actual *details* of the public API may =need fleshing out.

**Comment by elibarzilay** [comment]
@chriskrycho: so will you review it?  If not, then maybe it's best to get it in as is?

**Comment by chriskrycho** [comment]
Yes, I'm still going to review it; but we're not going to "get it in as is" without clear definition of what the actual public API is intended to be. :)

**Comment by chriskrycho** [comment]
@NullVoxPopuli can you please open *docs PRs* to `@ember/test-helpers`, and once those land we can get this through?

**Comment by elibarzilay** [comment]
@chriskrycho: if it's important to not get it in as-is, it might be a good idea to have some stub review that requests changes.  I'm just worried that in case you won't get to it today and forget about it, it'll be easy to get it in by mistake later.

**Comment by NullVoxPopuli** [comment]
Is there some rush that's going on here? :)

**Comment by elibarzilay** [comment]
No rush, it's on the DT maintainer queue so it's supposed to be ready for merging, and there's no clear sign that it's not (like a draft PR or a blocking review).  On Monday someone else will be on rotation, and getting it in can be a very easy mistake to do.

File: types/ember__test-helpers/index.d.ts
Comment (rwjblue) [inline types/ember__test-helpers/index.d.ts:201]: This isn't public API (accessing `owner.application`)

File: types/ember__test-helpers/index.d.ts
Comment (rwjblue) [inline types/ember__test-helpers/index.d.ts:195]: `Owner` should already have types, no?

Basically, this thing is "owner":

https://github.com/DefinitelyTyped/DefinitelyTyped/blob/15fe0696be464ec1f7cc97ed42f1ec35bc4476b5/types/ember__engine/instance.d.ts#L9

File: types/ember__test-helpers/index.d.ts
Comment (rwjblue) [inline types/ember__test-helpers/index.d.ts:196]: There are other APIs that are public and added to the context. From `@ember/test-helpers` perspective, these types refer to public APIs:

https://github.com/emberjs/ember-test-helpers/blob/dd5fbb1280274068c046a8ff0a86177e9507e5a1/addon-test-support/%40ember/test-helpers/setup-context.ts#L21-L31

As well as the `element` property (when running `setupRenderingTest`/`setupApplicationTest` variants, but not `setupTest`), that is defined in these locations:

* https://github.com/emberjs/ember-test-helpers/blob/dd5fbb1280274068c046a8ff0a86177e9507e5a1/addon-test-support/%40ember/test-helpers/setup-rendering-context.ts#L258

* https://github.com/emberjs/ember-test-helpers/blob/dd5fbb1280274068c046a8ff0a86177e9507e5a1/addon-test-support/%40ember/test-helpers/setup-application-context.ts#L134-L138

**Comment by chriskrycho** [comment]
Ah, @elibarzilay sorry, I wasn't realizing it needed a "blocking" review. Adding that now.

**Review by chriskrycho — CHANGES_REQUESTED** [review]
Thanks for the feedback on what is and isn't public here @rwjblue – exactly what we needed!

@NullVoxPopuli if you can submit a PR specifying these types a bit better in the API docs at ember-test-helpers, and update here per @rwjblue's comments on what is and is not public, that'd be awesome!

**Comment by elibarzilay** [comment]
@chriskrycho: actually I went off on a tangent, making the maintenance work easier in many cases, like this one...

**Pull request closed without merging**
